File: cpp_bind/id_utils.cpp
#include "id_utils.h"
#include "BulletInverseDynamics/MultiBodyTree.hpp"
#include "../Extras/InverseDynamics/MultiBodyNameMap.hpp"
#include "../Extras/InverseDynamics/MultiBodyTreeCreator.hpp"
#include "../Extras/InverseDynamics/btMultiBodyTreeCreator.hpp"
#include "../Extras/InverseDynamics/CloneTreeCreator.hpp"
#include "../Extras/InverseDynamics/SimpleTreeCreator.hpp"
#include "../Extras/InverseDynamics/User2InternalIndex.hpp"
#include "../Extras/InverseDynamics/IDRandomUtil.hpp"
#include "BulletDynamics/Featherstone/btMultiBody.h"
#include <cstring>

nkMultiBodyNameMapHandle nk_multibody_name_map_create()
{
    btInverseDynamics::MultiBodyNameMap* map = new btInverseDynamics::MultiBodyNameMap();
    return static_cast<nkMultiBodyNameMapHandle>(map);
}

void nk_multibody_name_map_destroy(nkMultiBodyNameMapHandle map)
{
    if (!map) return;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    delete nameMap;
}

int nk_multibody_name_map_add_body(nkMultiBodyNameMapHandle map, int index, const char* name)
{
    if (!map || !name) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    return nameMap->addBody(index, std::string(name));
}

int nk_multibody_name_map_add_joint(nkMultiBodyNameMapHandle map, int index, const char* name)
{
    if (!map || !name) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    return nameMap->addJoint(index, std::string(name));
}

int nk_multibody_name_map_get_body_name(nkMultiBodyNameMapHandle map, int index, char* out_name, int max_len)
{
    if (!map || !out_name || max_len <= 0) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    
    std::string name;
    int result = nameMap->getBodyName(index, &name);
    if (result == 0) {
        strncpy(out_name, name.c_str(), max_len - 1);
        out_name[max_len - 1] = '\0';
    }
    return result;
}

int nk_multibody_name_map_get_joint_name(nkMultiBodyNameMapHandle map, int index, char* out_name, int max_len)
{
    if (!map || !out_name || max_len <= 0) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    
    std::string name;
    int result = nameMap->getJointName(index, &name);
    if (result == 0) {
        strncpy(out_name, name.c_str(), max_len - 1);
        out_name[max_len - 1] = '\0';
    }
    return result;
}

int nk_multibody_name_map_get_body_index(nkMultiBodyNameMapHandle map, const char* name, int* out_index)
{
    if (!map || !name || !out_index) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    return nameMap->getBodyIndex(std::string(name), out_index);
}

int nk_multibody_name_map_get_joint_index(nkMultiBodyNameMapHandle map, const char* name, int* out_index)
{
    if (!map || !name || !out_index) return -1;
    btInverseDynamics::MultiBodyNameMap* nameMap = 
        static_cast<btInverseDynamics::MultiBodyNameMap*>(map);
    return nameMap->getJointIndex(std::string(name), out_index);
}

nkMultiBodyTreeCreatorHandle nk_multibody_tree_creator_from_bt_multibody(nkMultiBodyHandle btmb, int verbose)
{
    if (!btmb) return nullptr;
    
    btMultiBody* bt_multibody = static_cast<btMultiBody*>(btmb);
    
    btInverseDynamics::btMultiBodyTreeCreator* creator = 
        new btInverseDynamics::btMultiBodyTreeCreator();
    
    if (creator->createFromBtMultiBody(bt_multibody, verbose != 0) != 0) {
        delete creator;
        return nullptr;
    }
    
    return static_cast<nkMultiBodyTreeCreatorHandle>(creator);
}

void nk_multibody_tree_creator_destroy(nkMultiBodyTreeCreatorHandle creator)
{
    if (!creator) return;
    btInverseDynamics::MultiBodyTreeCreator* treeCreator = 
        static_cast<btInverseDynamics::MultiBodyTreeCreator*>(creator);
    delete treeCreator;
}

nkMultiBodyHandle nk_multibody_create_from_creator(nkMultiBodyTreeCreatorHandle creator)
{
    if (!creator) return nullptr;
    
    btInverseDynamics::MultiBodyTreeCreator* treeCreator = 
        static_cast<btInverseDynamics::MultiBodyTreeCreator*>(creator);
    
    btInverseDynamics::MultiBodyTree* tree = 
        btInverseDynamics::CreateMultiBodyTree(*treeCreator);
    
    return static_cast<nkMultiBodyHandle>(tree);
}

nkMultiBodyTreeCreatorHandle nk_clone_tree_creator_create(nkMultiBodyHandle reference)
{
    if (!reference) return nullptr;
    
    btInverseDynamics::MultiBodyTree* refTree = 
        static_cast<btInverseDynamics::MultiBodyTree*>(reference);
    
    btInverseDynamics::CloneTreeCreator* creator = 
        new btInverseDynamics::CloneTreeCreator(refTree);
    
    return static_cast<nkMultiBodyTreeCreatorHandle>(creator);
}

void nk_clone_tree_creator_destroy(nkMultiBodyTreeCreatorHandle creator)
{
    if (!creator) return;
    btInverseDynamics::CloneTreeCreator* cloneCreator = 
        static_cast<btInverseDynamics::CloneTreeCreator*>(creator);
    delete cloneCreator;
}

nkMultiBodyHandle nk_multibody_clone(nkMultiBodyHandle reference)
{
    if (!reference) return nullptr;
    
    btInverseDynamics::MultiBodyTree* refTree = 
        static_cast<btInverseDynamics::MultiBodyTree*>(reference);
    
    btInverseDynamics::CloneTreeCreator creator(refTree);
    btInverseDynamics::MultiBodyTree* clonedTree = 
        btInverseDynamics::CreateMultiBodyTree(creator);
    
    return static_cast<nkMultiBodyHandle>(clonedTree);
}

nkMultiBodyTreeCreatorHandle nk_simple_tree_creator_create(int num_bodies)
{
    if (num_bodies <= 0) return nullptr;
    
    btInverseDynamics::SimpleTreeCreator* creator = 
        new btInverseDynamics::SimpleTreeCreator(num_bodies);
    
    return static_cast<nkMultiBodyTreeCreatorHandle>(creator);
}

void nk_simple_tree_creator_destroy(nkMultiBodyTreeCreatorHandle creator)
{
    if (!creator) return;
    btInverseDynamics::SimpleTreeCreator* simpleCreator = 
        static_cast<btInverseDynamics::SimpleTreeCreator*>(creator);
    delete simpleCreator;
}

nkUser2InternalIndexHandle nk_user2internal_index_create()
{
    btInverseDynamics::User2InternalIndex* index = new btInverseDynamics::User2InternalIndex();
    return static_cast<nkUser2InternalIndexHandle>(index);
}

void nk_user2internal_index_destroy(nkUser2InternalIndexHandle handle)
{
    if (!handle) return;
    btInverseDynamics::User2InternalIndex* index = 
        static_cast<btInverseDynamics::User2InternalIndex*>(handle);
    delete index;
}

void nk_user2internal_index_add_body(nkUser2InternalIndexHandle handle, int body, int parent)
{
    if (!handle) return;
    btInverseDynamics::User2InternalIndex* index = 
        static_cast<btInverseDynamics::User2InternalIndex*>(handle);
    index->addBody(body, parent);
}

int nk_user2internal_index_build_mapping(nkUser2InternalIndexHandle handle)
{
    if (!handle) return -1;
    btInverseDynamics::User2InternalIndex* index = 
        static_cast<btInverseDynamics::User2InternalIndex*>(handle);
    return index->buildMapping();
}

int nk_user2internal_index_user2internal(nkUser2InternalIndexHandle handle, int user, int* out_internal)
{
    if (!handle || !out_internal) return -1;
    btInverseDynamics::User2InternalIndex* index = 
        static_cast<btInverseDynamics::User2InternalIndex*>(handle);
    return index->user2internal(user, out_internal);
}

int nk_user2internal_index_internal2user(nkUser2InternalIndexHandle handle, int internal, int* out_user)
{
    if (!handle || !out_user) return -1;
    btInverseDynamics::User2InternalIndex* index = 
        static_cast<btInverseDynamics::User2InternalIndex*>(handle);
    return index->internal2user(internal, out_user);
}

void nk_id_random_init()
{
    btInverseDynamics::randomInit();
}

void nk_id_random_init_with_seed(unsigned int seed)
{
    btInverseDynamics::randomInit(seed);
}

int nk_id_random_int(int low, int high)
{
    return btInverseDynamics::randomInt(low, high);
}

nkReal nk_id_random_float(nkReal low, nkReal high)
{
    return btInverseDynamics::randomFloat(low, high);
}

nkReal nk_id_random_mass()
{
    return btInverseDynamics::randomMass();
}

void nk_id_random_inertia_principal(nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!out_x || !out_y || !out_z) return;
    btInverseDynamics::vec3 inertia = btInverseDynamics::randomInertiaPrincipal();
    *out_x = inertia(0);
    *out_y = inertia(1);
    *out_z = inertia(2);
}

void nk_id_random_axis(nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!out_x || !out_y || !out_z) return;
    btInverseDynamics::vec3 axis = btInverseDynamics::randomAxis();
    *out_x = axis(0);
    *out_y = axis(1);
    *out_z = axis(2);
}

File: rust/src/core/id_utils.rs
use std::ffi::{c_void, CString, CStr};
use std::ptr::NonNull;

use super::inverse_dynamics::MultiBody;
use crate::ffi;

pub struct MultiBodyNameMap {
    handle: NonNull<c_void>,
}

impl MultiBodyNameMap {
    pub fn new() -> Self {
        let handle = unsafe { ffi::nk_multibody_name_map_create() };
        Self {
            handle: NonNull::new(handle).expect("Failed to create name map"),
        }
    }

    pub fn add_body(&self, index: i32, name: &str) -> i32 {
        let c_name = CString::new(name).unwrap_or_default();
        unsafe {
            ffi::nk_multibody_name_map_add_body(self.handle.as_ptr(), index, c_name.as_ptr())
        }
    }

    pub fn add_joint(&self, index: i32, name: &str) -> i32 {
        let c_name = CString::new(name).unwrap_or_default();
        unsafe {
            ffi::nk_multibody_name_map_add_joint(self.handle.as_ptr(), index, c_name.as_ptr())
        }
    }

    pub fn get_body_name(&self, index: i32) -> Option<String> {
        let mut buffer = [0i8; 256];
        let result = unsafe {
            ffi::nk_multibody_name_map_get_body_name(
                self.handle.as_ptr(),
                index,
                buffer.as_mut_ptr(),
                buffer.len() as i32,
            )
        };
        
        if result == 0 {
            unsafe {
                CStr::from_ptr(buffer.as_ptr())
                    .to_str()
                    .ok()
                    .map(|s| s.to_string())
            }
        } else {
            None
        }
    }

    pub fn get_joint_name(&self, index: i32) -> Option<String> {
        let mut buffer = [0i8; 256];
        let result = unsafe {
            ffi::nk_multibody_name_map_get_joint_name(
                self.handle.as_ptr(),
                index,
                buffer.as_mut_ptr(),
                buffer.len() as i32,
            )
        };
        
        if result == 0 {
            unsafe {
                CStr::from_ptr(buffer.as_ptr())
                    .to_str()
                    .ok()
                    .map(|s| s.to_string())
            }
        } else {
            None
        }
    }

    pub fn get_body_index(&self, name: &str) -> Option<i32> {
        let c_name = CString::new(name).ok()?;
        let mut index: i32 = -1;
        let result = unsafe {
            ffi::nk_multibody_name_map_get_body_index(
                self.handle.as_ptr(),
                c_name.as_ptr(),
                &mut index,
            )
        };
        
        if result == 0 { Some(index) } else { None }
    }

    pub fn get_joint_index(&self, name: &str) -> Option<i32> {
        let c_name = CString::new(name).ok()?;
        let mut index: i32 = -1;
        let result = unsafe {
            ffi::nk_multibody_name_map_get_joint_index(
                self.handle.as_ptr(),
                c_name.as_ptr(),
                &mut index,
            )
        };
        
        if result == 0 { Some(index) } else { None }
    }
}

impl Default for MultiBodyNameMap {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for MultiBodyNameMap {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_multibody_name_map_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for MultiBodyNameMap {}
unsafe impl Sync for MultiBodyNameMap {}

pub struct MultiBodyTreeCreator {
    handle: NonNull<c_void>,
}

impl MultiBodyTreeCreator {
    pub fn from_bt_multibody(btmb: &MultiBody, verbose: bool) -> Option<Self> {
        let handle = unsafe {
            ffi::nk_multibody_tree_creator_from_bt_multibody(
                btmb.handle(),
                if verbose { 1 } else { 0 },
            )
        };
        
        NonNull::new(handle).map(|h| Self { handle: h })
    }

    pub fn create_tree(&self) -> Option<MultiBody> {
        let handle = unsafe { ffi::nk_multibody_create_from_creator(self.handle.as_ptr()) };
        NonNull::new(handle).map(MultiBody::from_handle)
    }
}

impl Drop for MultiBodyTreeCreator {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_multibody_tree_creator_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for MultiBodyTreeCreator {}
unsafe impl Sync for MultiBodyTreeCreator {}

pub struct CloneTreeCreator {
    handle: NonNull<c_void>,
}

impl CloneTreeCreator {
    pub fn new(reference: &MultiBody) -> Option<Self> {
        let handle = unsafe { ffi::nk_clone_tree_creator_create(reference.handle()) };
        NonNull::new(handle).map(|h| Self { handle: h })
    }

    pub fn create_tree(&self) -> Option<MultiBody> {
        let handle = unsafe { ffi::nk_multibody_create_from_creator(self.handle.as_ptr()) };
        NonNull::new(handle).map(MultiBody::from_handle)
    }
}

impl Drop for CloneTreeCreator {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_clone_tree_creator_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for CloneTreeCreator {}
unsafe impl Sync for CloneTreeCreator {}

pub fn clone_multibody(reference: &MultiBody) -> Option<MultiBody> {
    let handle = unsafe { ffi::nk_multibody_clone(reference.handle()) };
    NonNull::new(handle).map(MultiBody::from_handle)
}

pub struct SimpleTreeCreator {
    handle: NonNull<c_void>,
}

impl SimpleTreeCreator {
    pub fn new(num_bodies: i32) -> Option<Self> {
        if num_bodies <= 0 {
            return None;
        }
        let handle = unsafe { ffi::nk_simple_tree_creator_create(num_bodies) };
        NonNull::new(handle).map(|h| Self { handle: h })
    }

    pub fn create_tree(&self) -> Option<MultiBody> {
        let handle = unsafe { ffi::nk_multibody_create_from_creator(self.handle.as_ptr()) };
        NonNull::new(handle).map(MultiBody::from_handle)
    }
}

impl Drop for SimpleTreeCreator {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_simple_tree_creator_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for SimpleTreeCreator {}
unsafe impl Sync for SimpleTreeCreator {}

pub struct User2InternalIndex {
    handle: NonNull<c_void>,
}

impl User2InternalIndex {
    pub fn new() -> Self {
        let handle = unsafe { ffi::nk_user2internal_index_create() };
        Self {
            handle: NonNull::new(handle).expect("Failed to create User2InternalIndex"),
        }
    }

    pub fn add_body(&self, body: i32, parent: i32) {
        unsafe {
            ffi::nk_user2internal_index_add_body(self.handle.as_ptr(), body, parent);
        }
    }

    pub fn build_mapping(&self) -> i32 {
        unsafe { ffi::nk_user2internal_index_build_mapping(self.handle.as_ptr()) }
    }

    pub fn user_to_internal(&self, user: i32) -> Option<i32> {
        let mut internal: i32 = -1;
        let result = unsafe {
            ffi::nk_user2internal_index_user2internal(self.handle.as_ptr(), user, &mut internal)
        };
        if result == 0 { Some(internal) } else { None }
    }

    pub fn internal_to_user(&self, internal: i32) -> Option<i32> {
        let mut user: i32 = -1;
        let result = unsafe {
            ffi::nk_user2internal_index_internal2user(self.handle.as_ptr(), internal, &mut user)
        };
        if result == 0 { Some(user) } else { None }
    }
}

impl Default for User2InternalIndex {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for User2InternalIndex {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_user2internal_index_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for User2InternalIndex {}
unsafe impl Sync for User2InternalIndex {}

pub mod random {
    use super::ffi;
    use crate::core::types::Real;

    pub fn init() {
        unsafe { ffi::nk_id_random_init() }
    }

    pub fn init_with_seed(seed: u32) {
        unsafe { ffi::nk_id_random_init_with_seed(seed) }
    }

    pub fn random_int(low: i32, high: i32) -> i32 {
        unsafe { ffi::nk_id_random_int(low, high) }
    }

    pub fn random_float(low: Real, high: Real) -> Real {
        unsafe { ffi::nk_id_random_float(low, high) }
    }

    pub fn random_mass() -> Real {
        unsafe { ffi::nk_id_random_mass() }
    }

    pub fn random_inertia_principal() -> (Real, Real, Real) {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_id_random_inertia_principal(&mut x, &mut y, &mut z);
        }
        (x, y, z)
    }

    pub fn random_axis() -> (Real, Real, Real) {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_id_random_axis(&mut x, &mut y, &mut z);
        }
        (x, y, z)
    }
}
